core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -675,11 +676,15 @@ private[spark] class TaskSchedulerImpl(
           // Check whether the barrier tasks are partially launched.
           // TODO SPARK-24818 handle the assert failure case (that can happen when some locality
           // requirements are not fulfilled, and we should revert the launched tasks).
-          require(addressesWithDescs.size == taskSet.numTasks,
-            s"Skip current round of resource offers for barrier stage ${taskSet.stageId} " +
```

Comment:
This is the primary issue causing the bug - and is overly strict.
Unfortunately, the way barrier scheduling has been written, the only way to get it working currently is to disable delay scheduling entirely.
We will need to relook how to get barrier scheduling working more gracefully with task locality and executor failures (and in future, dynamic allocation).

Reply:
Could we instead have a counter inside the taskSet or other mechanism to allow for X retries? It seems like turning it off bis a bit of a behaviour change from the point of view of considering backporting.

Reply:
proposal was to do that in a separate jira. https://issues.apache.org/jira/browse/SPARK-31509

Reply:
I agree with @holdenk.
`require` would have ended up throwing an exception in this case - we should do the same after taskSet.abort to prevent change in behavior - particularly for backport

Reply:
> Could we instead have a counter inside the taskSet or other mechanism to allow for X retries?

I believe barrier retry is next step we plan to do in the future release but not 2.4.

> It seems like turning it off bis a bit of a behaviour change from the point of view of considering backporting.

What's behavior change? Previously, application gets hang and now it fail as we expect in first place.

> require would have ended up throwing an exception in this case - we should do the same after taskSet.abort to prevent change in behavior - particularly for backport

To be honest, I'm fine to keep throwing exception there, but I disagree that throwing exception is expected behavior we can not change. Actually, no one would handle the exception thrown here. And I believe our expect behavior is to fail the application with the clear error message.

Reply:
@Ngone51 It is not whether there is user code is handling exception or not - but throwing exception stops the flow of code here. If exception is not thrown, rest of the code will execute - barrier initialization, flag updates, log messages, etc.
Some of it gets cleaned up by taskset abort, some of it results in confusing log messages.

Reply:
I see, make sense.

Reply:
I also think that throw exception isn't recommended way to stop code flow within the scope of rpc. The better way is to rewrite the code logic and gets every branch under control.

Reply:
I definitely agree !
I was referring specifically in context of backport - not throwing an exception results in behavior change, which we should not have.
Ideal solution is to fix this - the `require` was a precondition we were expecting to be satisfied for barrier scheduling - which gets violated when locality preference is present (which probably the original authors did not expect/consider).